### PR TITLE
Add knn-image-classifier

### PR DESCRIPTION
**Pull request by HalfdanJ**
This is a port of the older deeplearnjs implementation. Fixing https://github.com/tensorflow/tfjs/issues/385. 

The classifier has been ported to use mobilenet instead of squeezenet. The demo uses the code from teachable-machine-boilerplate.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/34)

### CODE REVIEW

**Comment by oveddan** [comment]
Review status: 0 of 1 LGTMs obtained

---

*[knn-image-classifier/src/index.ts, line 22 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cpUkECKf1Mm3MeWT:-LF4cpUkECKf1Mm3MeWU:b1leoe) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L22)):*
> ```XML
> import {Tensor1D, Tensor2D, Tensor3D} from '@tensorflow/tfjs';
> 
> class KNNImageClassifier {
> ```

Could you add a tsdoc description of what the class does?

---

*[knn-image-classifier/src/index.ts, line 47 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cvFH1RCshRPtaiVO:-LF4cvFH1RCshRPtaiVP:bps69yb) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L47)):*
> ```XML
>   clearClass(classIndex: number) {
>     if (classIndex >= this.numClasses) {
>       console.log('Cannot clear invalid class ${classIndex}');
> ```

Throw an error here?

---

*[knn-image-classifier/src/index.ts, line 69 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cZR1BiPP5Y9odMNI:-LF4cZR1BiPP5Y9odMNJ:b-n8dc59) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L69)):*
> ```XML
>       // logits matrix.
>       const _logits = this.net.infer(image).as1D();
>       const logits = this.normalizeVector(_logits);
> ```

These two should have better names, maybe something like:
const logits = ....
const normalizedLogits = ...

or 

const logits = this.normalizeVector(this.net.infer(image).as1D());

---

*[knn-image-classifier/src/index.ts, line 73 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cjkrCQuwZ-S4C2_d:-LF4cjks2uYFL0HF-J81:bsosu29) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L73)):*
> ```XML
>       const logitsSize = logits.shape[0];
>       // console.log(logitsSize);
>       console.log(logits, logitsSize, logits.shape);
> ```

Remove above two lines

---

*[knn-image-classifier/src/index.ts, line 110 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4emAUESTais3517uw:-LF4emAUESTais3517ux:b7iwuec) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L110)):*
> ```XML
>     return tf.tidy(() => {
>       const _logits = this.net.infer(image).as1D();
>       const logits = this.normalizeVector(_logits);
> ```

These same two lines appear in `addImage`, extract into a function `this.inferNormalizedLogits(image)`

---

*[knn-image-classifier/src/index.ts, line 133 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4dNr56wV8cszto4Mp:-LF4dNr56wV8cszto4Mq:b-3etjc6) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L133)):*
> ```XML
> 
>       const numExamples = this.getNumExamples();
>       return this.trainLogitsMatrix.as2D(numExamples, logitsSize)
> ```

Not exactly clear what this is doing...maybe create a function with a name that explains what its returning?

something like:

return convertToBlah(this.trainLogitsMatrix);

---

*[knn-image-classifier/src/index.ts, line 147 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4dqCN5iSvxYq2kEZ9:-LF4dqCN5iSvxYq2kEZA:b-4u55n3) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L147)):*
> ```XML
>    * values for all possible classes.
>    */
>   async predictClass(image: Tensor3D):
> ```

This is a bit of a long function.  Can it be broken down into smaller parts?

---

*[knn-image-classifier/src/index.ts, line 223 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4e1B4-6kUqWqVkA2g:-LF4e1B4-6kUqWqVkA2h:bteuydn) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L223)):*
> ```XML
>   }
> 
>   private concatWithNulls(ndarray1: Tensor2D, ndarray2: Tensor2D): Tensor2D {
> ```

This should be moved to static utility function outside of the class, since it does not refer to anything that is part of the class.  It would also make it more testable.

---

*[knn-image-classifier/src/index.ts, line 238 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4eDn-BwKDvHevWnnF:-LF4eDn05n0hs8R2ga-u:bpt0lt1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L238)):*
> ```XML
>    * Normalize the provided vector to unit length.
>    */
>   private normalizeVector(vec: Tensor1D) {
> ```

Same as above - This should be moved to static utility function.

---

*[knn-image-classifier/src/index.ts, line 259 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4eK6N7V6gdbJtCELa:-LF4eK6N7V6gdbJtCELb:b-jesi1w) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L259)):*
> ```XML
>   }
> 
>   private topK(values: Float32Array, k: number):
> ```

Same as above - This should be moved to a static utility function. 

---

*[knn-image-classifier/src/index.ts, line 285 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4g24k-T5Q1kDtTfFE:-LF4g24k-T5Q1kDtTfFF:b4hlrkb) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L285)):*
> ```XML
> }
> 
> async function load(numClasses: number, topK: number) {
> ```

Can you add tsdoc comments to the signature?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/34)*

**Comment by HalfdanJ** [comment]
Review status: 0 of 1 LGTMs obtained

---

*[knn-image-classifier/src/index.ts, line 22 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cpUkECKf1Mm3MeWT:-LFIAmflENx7WeyOGnlK:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L22)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Could you add a tsdoc description of what the class does?
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 47 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cvFH1RCshRPtaiVO:-LFIAnflBW6CCUHqMW4y:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L47)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Throw an error here?
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 69 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cZR1BiPP5Y9odMNI:-LFIC4km96VbhV06_aLF:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L69)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

These two should have better names, maybe something like:
const logits = ....
const normalizedLogits = ...

or 

const logits = this.normalizeVector(this.net.infer(image).as1D());
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 73 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4cjkrCQuwZ-S4C2_d:-LFIC6di97GPcJjAbYJX:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L73)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Remove above two lines
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 110 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4emAUESTais3517uw:-LFIC7jM5HMEvDYSIWGx:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L110)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

These same two lines appear in `addImage`, extract into a function `this.inferNormalizedLogits(image)`
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 133 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4dNr56wV8cszto4Mp:-LFICGez6KKUkb2qVhZ5:bwyh7th) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L133)):*
> ```
> mExamples = this.getNumEx
> ```
 Also not entirely sure... Its inherited from tfjs-core  https://github.com/tensorflow/tfjs-core/blob/master/models/knn_image_classifier/knn_image_classifier.ts#L155

---

*[knn-image-classifier/src/index.ts, line 147 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4dqCN5iSvxYq2kEZ9:-LFIGgOv2OKjtiGZnCb6:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L147)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

This is a bit of a long function.  Can it be broken down into smaller parts?
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 223 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4e1B4-6kUqWqVkA2g:-LFII5pb7DHV-4StA0sE:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L223)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

This should be moved to static utility function outside of the class, since it does not refer to anything that is part of the class.  It would also make it more testable.
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 238 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4eDn-BwKDvHevWnnF:-LFIHwQy0cUqUzsNkxZ5:bdvx3pq) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L238)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Same as above - This should be moved to static utility function.
</blockquote></details>

This function has a scalar that is kept over time, and disposed in dispose. Not sure how to move it out to a static function, and maintain that variable

---

*[knn-image-classifier/src/index.ts, line 259 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4eK6N7V6gdbJtCELa:-LFIJ6tW4ykKItytaV23:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L259)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Same as above - This should be moved to a static utility function. 
</blockquote></details>

Done.

---

*[knn-image-classifier/src/index.ts, line 285 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-LF4g24k-T5Q1kDtTfFE:-LFIJXRTBkNeoYuJQUJ1:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/e9256dfc9e6dba9fe94c418e063186a929402741/knn-image-classifier/src/index.ts#L285)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Can you add tsdoc comments to the signature?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/34)*

**Comment by nsthorat** [comment]
:lgtm_strong:

Thanks for doing this! I will take on phase 2 of generalizing this to any embedding.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/34#-:-LFIO_SlEjrRaogi53Vm:bpr2pbb)*
